wgengine/magicsock/magicsock.go

```diff
@@ -355,55 +360,229 @@ func (c *Conn) LocalPort() uint16 {
 	return uint16(laddr.Port)
 }
 
-func (c *Conn) Send(b []byte, ep device.Endpoint) error {
-	a := ep.(*AddrSet)
-
+func shouldSprayPacket(b []byte) bool {
+	if len(b) < 4 {
+		return false
+	}
 	msgType := binary.LittleEndian.Uint32(b[:4])
 	switch msgType {
-	case device.MessageInitiationType, device.MessageResponseType, device.MessageCookieReplyType:
-		// Part of the wireguard handshake.
-		// Send to every potential endpoint we have for a peer.
-		a.mu.Lock()
-		roamAddr := a.roamAddr
-		a.mu.Unlock()
-
-		var err error
-		var success bool
-		if roamAddr != nil {
-			_, err = c.pconn.WriteTo(b, roamAddr)
-			if err == nil {
-				success = true
-			}
+	case device.MessageInitiationType,
+		device.MessageResponseType,
+		device.MessageCookieReplyType: // TODO: necessary?
+		return true
+	}
+	return false
+}
+
+// packetDsts returns the destinations that b should be written to in
+// order to reach as. Some of the returned UDPAddrs may be fake addrs
+// representing DERP servers.
+//
+// If oneDst is non-nil, dsts should be ignored. The oneDsts is only
+// returned if there's exactly 1 destination, as there is in the common
+// fast path. This avoids an allocation.
```

Comment:
Might be able to avoid all allocations with `appendPacketDsts(as *AddrSet, b []byte, dsts []*net.UDPAddr) []*net.UDPAddr`.

The `Send` method should be able to keep a slice that it passes on the stack.

Reply:
I changed it to append for style reasons (I didn't like oneDest), but I haven't profiled either way yet.

I'll do that in its own PR, with a Benchmark func.